### PR TITLE
STS tests based on constraints

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

This PR introduces the test setup for testing STS rules using the constraint library. It's a WIP effort but we have a running test for a simple STS rule: GOVCERT.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/cardano-ledger-conway.cabal [hunk 1]
```diff
@@ -56,7 +57,6 @@ library
         Cardano.Ledger.Conway.Rules.Tickf
         Cardano.Ledger.Conway.Rules.Utxos
         Cardano.Ledger.Conway.Rules.Utxow
-        Cardano.Ledger.Conway.TxAuxData
```
Comment (lehins) [inline eras/conway/impl/cardano-ledger-conway.cabal]: Why the move? It breaks the alphabetical order

File: libs/cardano-ledger-test/cardano-ledger-test.cabal [hunk 2]
```diff
@@ -138,7 +139,8 @@ library
         time,
         transformers,
         unordered-containers,
-        vector
+        vector,
+        lens
```
Comment (lehins) [inline libs/cardano-ledger-test/cardano-ledger-test.cabal]: We don't use `lens`. We use `microlens` on this project
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+genUniverse = generate (genSubstFromConstraints (Conway Standard) standardOrderInfo (universePreds def (Conway Standard)))
+
+-- NOTE: these tests can be slow (~10 seconds) because it takes time to generate
+-- the initial universe. Reducing the number of tests will NOT help.
+tests_STS :: TestTree
+tests_STS =
+  withResource genUniverse (const $ pure ()) $ \ io_subst ->
+  testGroup "STS property tests"
+    [ testProperty "prop_GOVCERT" $ ioProperty (prop_GOVCERT <$> io_subst)
+    ]
```
Comment (MaximilianAlgehed) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: @TimSheard this is where we re-use the universe we generate. The nice thing here is that `io_subst` can be re-used in all the different properties.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: I don't see any instances in this module
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+module Test.Cardano.Ledger.STS where
+
+import Control.Lens ((^.))
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]:
```suggestion
import Lens.Micro ((^.))
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    , Right v <- runTyped (runTerm env tm) ->
+        Just (Name x, Payload rep v acc)
+  _ -> Nothing
+  where
+    known = isJust . flip findName env
+    unknown = not . known
+    knownTerm :: forall t. Term era t -> Bool
+    knownTerm = all known . varsOfTerm mempty
+    knownSums :: forall r. [Sum era r] -> Bool
+    knownSums = all known . foldl varsOfSum mempty
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: Avoiding space leaks even in tests is a good idea :slightly_smiling_face: 
```suggestion
    knownSums = all known . foldl' varsOfSum mempty
```

**Review by lehins — CHANGES_REQUESTED** [review]
That's a good start with STS. Only minor changes are needed.

File: eras/conway/impl/cardano-ledger-conway.cabal (hunk 1 above)
Comment (MaximilianAlgehed) [reply eras/conway/impl/cardano-ledger-conway.cabal]: Whoops. Probably a rebase artifact.

**Comment by MaximilianAlgehed** [comment]
@lehins I think this should be ready to merge

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  case result of
+    Left errs -> error $ unlines errs
+    Right subst -> pure subst
+
+shrinkFromConstraints :: Era era => OrderInfo -> [Pred era] -> RootTarget era t t -> t -> [t]
+shrinkFromConstraints order cs target val = do
+  let env = saturateEnv (getTarget val target emptyEnv) cs
+  graph <- monadTyped $ compile order $ rewrite cs
+  env' <- shrinkEnv graph env
+  monadTyped $ runTarget env' target
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: Be nice if we had some comments here, something like
  let env = saturateEnv (getTarget val target emptyEnv) cs
  -- Build a new env, with just the variables in the target, where the vars are bound to the values extracted from 'val'
 graph <- monadTyped $ compile order $ rewrite cs
-- Build a new graph from the original constraints
  env' <- shrinkEnv graph env
-- Shrink each of the values bound in the env, by using the graph to determine what other variables they depended on
 monadTyped $ runTarget env' target
-- Rebuild the 'val' using the target and the new env with the shrunk  values.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+saturateEnv env0 preds = go env0 preds
+  where
+    go env [] = env
+    go env (p : ps)
+      | Just (x, v) <- solveUnknown env p = saturateEnv (storeName x v env) preds
+      | otherwise = go env ps
+
+solveUnknown :: forall era. Era era => Env era -> Pred era -> Maybe (Name era, Payload era)
+solveUnknown env p = case p of
+  SumsTo _ (Var x@(V _ rep acc)) EQL sums
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: lots of missing cases when we pattern match against 'p'.  Do we have a plan to cover these as well?

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        Left errors -> counterexample ("lookupTerm: " ++ show errors) False
+        Right val -> property $ cont val
+
+------------------------------------------------------------------------
+-- Properties
+------------------------------------------------------------------------
+
+prop_GOVCERT :: Subst (ConwayEra StandardCrypto) -> Property
+prop_GOVCERT sub =
+  forAllShrink (pParamsGen sub) (pParamsShrink sub) $ \(unPParams -> pParams) ->
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: Try the DELEG, POOL, or UTXO rule. They will have some interesting properties

**Review by TimSheard — COMMENTED** [review]
The goal of this PR is to prototype the use of the shrinking capabilities by testing that a rules maintain the constraints we used to generates a 'legitimate' state. We should add another rule, as suggested in the comments, where there are some real constraints. This way we get a better test of the capabilities.

Personally, I'd like to better understand how we shrink an environment using the constraints

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs (hunk 8 above)
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: Yes, my plan is to implement these as they start to appear.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs (hunk 9 above)
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: Yes, I have them in my todo file :)

**Review by lehins — APPROVED** [review]
:+1: Thank you

**Comment by MaximilianAlgehed** [comment]
@TimSheard I think I've addressed your comments

**Comment by MaximilianAlgehed** [comment]
@lehins I added the `POOL` and `DELEG` rules here too. It might be worth while to merge this soon to avoid things changing beneath our feet.

**Pull request merged**